add pwm driver and example

conorpp opened the pull request:
PWM driver.  Tested works well on Solo2 prototype.

nickray commented on examples/pwm.rs:
```diff
@@ -0,0 +1,104 @@
+};
+pub use hal::typestates::pin::state;
+
+// translated from https://stackoverflow.com/a/2284929/2490057
```
An alternative could be https://docs.rs/micromath/, not sure if it's overkill?

nickray replied:
Specifically, https://github.com/NeoBirth/micromath/blob/develop/src/f32ext/sin.rs#L7-L9 which uses https://github.com/NeoBirth/micromath/blob/develop/src/f32ext/cos.rs#L9-L15

conorpp replied:
I think for the example it's fine to keep it more standalone.  For actual application probably makes sense to use micromath or something similar.

conorpp commented on src/typestates/pin.rs:
```diff
@@ -65,8 +65,17 @@ pub mod state {
     }
 
+    #[derive(Copy,Clone)]
+    pub enum Channel {
```
I'll rename Channel to CtimerMatchChannel.  

How other special functions are handled I think depends too much on the function.  For PWM & ADC functions, there are "channels" that need to be used by application later.  For IP dedicated pins (MOSI,MISO,SCA,SCL, etc) they are more "configure and forget", which can use the `Special<function::>` state.

For the ADC peripheral, it can do an "arduino-style" read using the Analog pin reference.

```rust
impl Adc
{
    // ...
    pub fn read(&mut self, pin: & Pin<impl PinId, state::Analog<direction::Input>>) -> Result<u16> {
        self.arm_normal_channel(pin.state.channel);
        //...
```

For the PWM peripheral, I'm implemented the wg trait, and I don't think there is a way to use the `Pin` type.

```rust
impl<TIMER> wg::Pwm for Pwm<TIMER>
where TIMER: Ctimer<init_state::Enabled>
{
    // error: "`impl Trait` in type aliases is unstable".
    type Channel = Pin<impl PinId, CtimerMatchChannel>;
```

I want the channel to be tied to the Pin type though, so adding a `channel() -> CtimerMatchChannel` method was the next best thing I could think of.  

Adding a local typedef means the user would need to manually match the correct typedef with the pin that was configured. E.g. 

```rust
let led_pin = pins.pio1_4.into_ctimer2_mat1(&mut iocon);
pwm.set_duty(Pwm::Channels::Channel0, 55)
```

vs:

```rust
let led_pin = pins.pio1_4.into_ctimer2_mat1(&mut iocon);
pwm.set_duty(led_pin.channel() /* -> Channel0 */, 55)
```

conorpp commented:
Conflict in `touch.rs` so had to make the merge commit I believe.

I renamed `macro_rules! cmat_pins` to `macro_rules! ctimer_match_output_pins` and renamed `Channel` to `CtimerMatchChannel`.  One discussion unresolved.

nickray commented:
If there's a conflict, you need to resolve it in the rebase of your feature onto main (starting with `git reset --hard HEAD~1` to remove the merge commit).

By not doing this, you're putting the onus of figuring out what's going on on me as reviewer now, and anyone who tries to follow along in the future too. (git does not always "do the right thing")

conorpp commented:
Thanks for the tips!  I fixed it without using a merge commit.

conorpp commented:
After using my prior PWM abstract a bit, I decided to revise it and am now up using the `Special<function::CTIMER_MAT>>` approach.
